File: OJ practices/LeetCode题目集/55. Jump Game(medium).java
/*
 * @Author: SourDumplings
 * @Date: 2020-03-25 20:16:54
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode-cn.com/problems/jump-game/
 * 
 * 使用贪心算法，每次都跳到能让下一跳最远的位置处，然后往回扫描
 */

class Solution
{
    public boolean canJump(int[] nums)
    {
        int l = nums.length;
        if (l == 1)
        {
            return true;
        }
        int thisStart = 0;
        int maxNextJumpD = 0;
        int lastStart = 0;
        while (thisStart < l - 1)
        {
            boolean flag = false;
            for (int j = maxNextJumpD; lastStart <= j; --j)
            {
                if (j + nums[j] > maxNextJumpD)
                {
                    maxNextJumpD = j + nums[j];
                    flag = true;
                }
            }
            if (!flag)
            {
                return false;
            }
            lastStart = thisStart;
            thisStart = maxNextJumpD;
        }
        return true;
    }
}

File: OJ practices/LeetCode题目集/55. Jump Game(medium).cpp
/**
 * @file 55. Jump Game(medium).cpp
 * @author chadchang (chadchang@tencent.com)
 * @brief 判断能否跳到目的地
 * @version 1.0.0
 * @date 2024-03-20
 * 
 * @copyright Copyright (c) 2024 SourDumplings
 * 
 * 贪心算法
 * 看起点到能跳最远处的点之间的点谁能再跳到最远，作为下一次跳的起点
 * 复杂度：时间 O(n) 空间 O(1)
 * 
 * https://leetcode.cn/problems/jump-game/description/?envType=study-plan-v2&envId=top-interview-150
 */

class Solution
{
public:
    bool canJump(vector<int> &nums)
    {
        int n = nums.size();
        int maxReach = nums[0];
        int s = 0;
        while (maxReach < n - 1)
        {
            int nextS = s;
            int nextMaxReach = maxReach;
            for (int i = s + 1; i <= maxReach; ++i)
            {
                if (nextMaxReach < nums[i] + i)
                {
                    nextMaxReach = nums[i] + i;
                    nextS = i;
                }
            }
            if (s == nextS)
            {
                return false;
            }
            
            s = nextS;
            maxReach = nextMaxReach;
        }
        return true;
    }
};